### DISCUSSION TITLE
GPU acceleration doesn't work with SelfHost BrowserSubProcess in CefSharp 121

### DISCUSSION THREAD

**Discussion by n303ef** [cefsharp/CefSharp#4707] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

121.3.4

### Operating System

Windows 11

### Architecture

x64

### .Net Version

.Net 4.7.2

### Implementation

WinForms

### Reproduction Steps

Enable SelfHost BrowserSubProcess (or enable Chrome Runtime which automatically enables Self Host) and open: [https://get.webgl.org/](https://get.webgl.org/)

### Expected behavior

Should confirm WebGL support and show 3D shape image.

### Actual behavior

You can see text "Hmm. While your browser seems to support WebGL, it is disabled or unavailable."

### Regression?

This worked in CefSharp 120 and lower versions

### Known Workarounds

Browser surely is created with `WebGl =CefState.Enabled`
Problem doesn't reproduce in "cefclient.exe --multi-threaded-message-loop --no-sandbox --enable-chrome-runtime"

### Does this problem also occur in the CEF Sample Application

No

### Other information

It works if we change renderer to GLES:

`settings.CefCommandLineArgs.Add("use-angle", "gles");`

**Comment by omerbalash** [comment]
This is not a CefSharp exclusive issue. I am seeing it in C++ as well. My GPU process has the exact same arguments as the cefclient one, and i see the EGL libraries are loaded into the process. I tried setting everything to default similar to the cefclient sample and setting webgl to enabled explicitly during browser creation. But the issue is reproduced. 

If i choose Swiftshader as webgl it does work however. If you find a solution could you please update the thread?

**Comment by n303ef** [comment]
> This is not a CefSharp exclusive issue

But why is it working well without SelfHost?

**Comment by omerbalash** [comment]
> > This is not a CefSharp exclusive issue
> 
> But why is it working well without SelfHost?

I am not sure. There isnt a SelfHost equivalent in C++. Moreover it looks like using the switch "use-angle" allows to also select both EGL and GLES as a workaround. it looks like in certain conditions the default value doesnt work.

**Comment by n303ef** [comment]
Log entries:

```
[0207/103431.898:ERROR:command_buffer_proxy_impl.cc(131)] ContextResult::kTransientFailure: Failed to send GpuControl.CreateCommandBuffer.
[0207/103431.904:ERROR:gpu_process_host.cc(992)] GPU process exited unexpectedly: exit_code=-2147483645
```

It seems to be GPU process crash when trying to start default renderer and then browser switches to software renderer.

**Comment by n303ef** [comment]
I've found solution. We should add app.manifest and enable correct values in "application" section under "compatibility" section.

**Comment by helderjgoncalves** [comment]
I have a similar issue, if not the same.
```
ERROR:gpu_process_host.cc(953)] GPU process exited unexpectedly: exit_code=-2147483645
```

What did you add to the `app.manifest` to solve the issue @n303ef ?
Currently, I only have the `<supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}" />`  tag.. (Windows 10/11).

Temporary fix: `--use-angle=d3d12` or `--use-angle=d3d9`. 
Not working: `--use-angle=d3d11`

-----------------------
Also found GPU/Vulkan issues on macOS..

**Reply by amaitland** [reply]
> Also found GPU/Vulkan issues on macOS..

`CefSharp` is windows only, so my guess is that you're not actually using `CefSharp`.

General `CEF` support forum is at https://www.magpcss.org/ceforum/index.php

**Comment by amaitland** [comment]
Official self hosted documentation is at https://github.com/cefsharp/CefSharp/wiki/SelfHost-BrowserSubProcess

The `MinimalExample` has example `app.manifest` files for different `WinForms/WPF/OffScreen` implementations.

https://github.com/cefsharp/CefSharp.MinimalExample
